pkg/cloudprovider/aws/aws.go

```diff
@@ -601,3 +614,42 @@ func addASGTags(config *cloudprovider.NodeGroupConfig, asg *autoscaling.Group, p
 		log.Errorf("failed to create auto scaling tag for ASG %v", id)
 	}
 }
+
+// terminateOrphanedInstances will attempt to terminate a list of instances
+func terminateOrphanedInstances(n *NodeGroup, instances []*string) {
```

Comment:
Oh further inspection, it looks like instances slice may not be all the instances in the ASG at all times. Looking at the batch logic, instances slice is mutated and any calls after the final call will be a nil slice. So the call to terminateInstances will be ineffectual.

Reply:
I'll add some tests in node_group_test.go for this

Reply:
After thinking about this for a bit, I'm not sure the best way to implement these test cases. As it's written right now, `terminateOrphanedInstances` is just a best-effort termination, so there's not an easy way to verify whether the correct number of instances are terminated, maybe you can elaborate on what you had in mind?

Reply:
Mainly thinking just to have a few cases that verify the correct number of instances were attempted to be terminated depending on the reason we're terminating them.
For example, in `setASGDesiredSizeOneShot` if the `case <-deadline.C:` is reached, it tries to terminate all instances returned from the fleet (so the test would expect N instances were asked to be terminated in the API), however in the batch loop `n.provider.service.AttachInstances` fails and we only ask the API terminate to terminate 20 (batchSize).